test/commands/source/retrieve.test.ts

```diff
@@ -266,4 +271,34 @@ describe('force:source:retrieve', () => {
     expect(displayStub.calledOnce).to.equal(false);
     expect(getJsonStub.calledOnce).to.equal(true);
   });
+
+  it('should warn users when retrieving CustomField with --metadata', async () => {
+    const metadata = 'CustomField:CustomObject__c.CustomField__c';
+    buildComponentSetStub.restore();
+    buildComponentSetStub = stubMethod(sandbox, ComponentSetBuilder, 'build').resolves({
+      retrieve: retrieveStub,
+      getPackageXml: () => packageXml,
+      toArray: () => {
+        return [exampleCustomFieldSourceComponent];
+      },
+    });
+    await runRetrieveCmd(['--metadata', metadata]);
+    expect(warnStub.calledOnce);
+    expect(warnStub.firstCall.firstArg).to.equal(messages.getMessage('wantsToRetrieveCustomFields'));
+  });
+
+  it('should warn users when retrieving CustomField with --manifest', async () => {
+    const manifest = 'package.xml';
```

Comment:
package.xml should look something like a real package.xml.  Or a NUT might be easier (--json will have a warning property)

<types>
  <members>*
  <CustomField>
</types>

Reply:
Since we're interested in the `componentSet` array I tought It would be a good idea to stub `ComponentSetBuilder.build` with different reponses (given real examples) for both wildcard options as well as `CustomField,CustomObject` I think that should cover the cases where the warn should and should not be displayed.